app-libs/exchange-oracle/src/coingecko.rs

```diff
@@ -0,0 +1,148 @@
+use serde::{Deserialize, Serialize};
+use std::{
+	string::{String, ToString},
+	time::Duration,
+	vec::Vec,
+};
+use url::Url;
+
+const COINGECKO_URL: &str = "https://api.coingecko.com";
+const COINGECKO_PARAM_CURRENCY: &str = "vs_currency";
+const COINGECKO_PARAM_COIN: &str = "ids";
+const COINGECKO_PATH: &str = "api/v3/coins/markets";
+
+/// REST client to make requests to CoinGecko.
+pub struct CoinGeckoClient {
+	client: RestClient<HttpClient>,
+}
+impl CoinGeckoClient {
+	pub fn new(baseurl: Url) -> Self {
+		let http_client = HttpClient::new(true, Some(Duration::from_secs(3u64)), None, None);
+		let rest_client = RestClient::new(http_client, baseurl);
+		CoinGeckoClient { client: rest_client }
+	}
+	pub fn base_url() -> Result<Url, Error> {
+		Url::parse(COINGECKO_URL).map_err(|e| Error::Other(format!("{:?}", e).into()))
+	}
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct CoinGeckoMarketStruct {
+	id: String,
+	symbol: String,
+	name: String,
+	current_price: Option<f32>,
+	last_updated: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct CoinGeckoMarket(pub Vec<CoinGeckoMarketStruct>);
+
+impl RestPath<String> for CoinGeckoMarket {
+	fn get_path(path: String) -> Result<String, itc_rest_client::error::Error> {
+		Ok(path)
+	}
+}
+
+impl GetExchangeRate for CoinGeckoClient {
+	fn get_exchange_rate(&mut self, coin: &str, vs_currency: &str) -> Result<f32, Error> {
+		let response = self
+			.client
+			.get_with::<String, CoinGeckoMarket>(
+				COINGECKO_PATH.to_string(),
+				&[(COINGECKO_PARAM_CURRENCY, vs_currency), (COINGECKO_PARAM_COIN, coin)],
+			)
+			.map_err(RestClientError)?;
+		let list = response.0;
+		if list.is_empty() {
+			error!("Got no market data from coinGecko. Check params {},{}", vs_currency, coin);
+			return Err(Error::NoValidData)
+		}
+		match list[0].current_price {
+			Some(r) => Ok(r),
+			None => {
+				error!("Failed to get the exchange rate of {} to {}", vs_currency, coin);
+				Err(Error::NoneExchangeRate)
+			},
+		}
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+
+	#[test]
+	fn get_exchange_rate_for_defined_coins_to_usd_from_coingecko_works() {
+		check_exchange_rate_is_existing_for("polkadot", "usd");
+
+		check_exchange_rate_is_existing_for("bitcoin", "usd");
+
+		check_exchange_rate_is_existing_for("ethereum", "usd");
+	}
+
+	#[test]
+	fn get_exchange_rate_for_defined_coins_to_chf_from_coingecko_works() {
+		check_exchange_rate_is_existing_for("polkadot", "chf");
+
+		check_exchange_rate_is_existing_for("bitcoin", "chf");
+
+		check_exchange_rate_is_existing_for("ethereum", "chf");
+	}
+
+	#[test]
+	fn get_exchange_rate_for_undefined_coins_fails() {
+		let url = CoinGeckoClient::base_url().unwrap();
+		let mut coingecko_client = CoinGeckoClient::new(url);
+		let result = coingecko_client.get_exchange_rate("dot", "usd");
+		let error = result.unwrap_err();
+		if let Error::NoValidData = error {
+		} else {
+			assert!(false, "Invalid error kind: {:?}", error);
+		}
```

Comment:
To check for a specific result/error, you can use the `assert_matches!` macro (needs to be enabled at the top of `lib.rs` with
`#![feature(assert_matches)]`). Then you can simply write:
```
assert_matches!(result, Err(Error::NoValidData));
```